File: src/ScalaSolutions/src/medium/LongestSubstring.scala
package medium

object LongestSubstring extends App {
  def lengthOfLongestSubstring(s: String): Int =
    s.foldRight(List.empty[Char])((x, acc) =>
      if (acc.contains(x)) acc else x :: acc
    ).length

  /** Find the longest continuous substring `l` without repeating characters.
    *
    * @param s
    *   String
    * @return
    *   The result lenght.
    */
  def longestSubstring(s: String): String = {
    def go(
        input: List[Char],
        stack: List[Char]
    ): String = (input, stack) match {
      case (Nil, Nil) => ""
      case (Nil, _)   => ""
      case (x :: xs, stack) if !stack.contains(x) =>
        x.toString + longestSubstring(go(xs, x :: stack).mkString)
      case (_, _) => ""
    }

    go(s.toList, Nil).mkString
  }

  /** Using an array to store `ASCII` character indexes. ChatGPT solution.
    */
  def lengthOfLongestSubstringG(s: String): Int = {
    val n = s.length
    var maxLength = 0
    var start = 0
    val charIndex =
      Array.fill(256)(-1)

    for (end <- 0 until n) {
      val char = s(end)
      if (charIndex(char) >= start) {
        start = charIndex(char) + 1
      }
      charIndex(char) = end
      maxLength = Math.max(maxLength, end - start + 1)
    }

    maxLength
  }

  println(lengthOfLongestSubstringG("abcaagbf"))
}

File: src/HaskellSolutions/src/Medium/LongestSubstring.hs
module Medium.LongestSubstring where

import qualified Data.Map as Map

longestSubstring' :: [Char] -> [Char]
longestSubstring' [] = []
longestSubstring' l = go l []
  where
    go [] [] = []
    go [] _ = []
    go (x:xs) stack
      | x `notElem` stack = x : longestSubstring' (go xs (x : stack))
      | otherwise = []

-- | Find the length of the longest substring `l` without repeating characters
longestSubstring :: [Char] -> Int
longestSubstring = length . longestSubstring'

-- | Uses the two-pointer algorithm. ChatGPT generated.
lengthOfLongestSubstring :: String -> Int
lengthOfLongestSubstring s = go 0 0 0 Map.empty
  where
    go start end maxLength charMap
      | end == length s = maxLength
      | otherwise =
        let currentChar = s !! end
         in case Map.lookup currentChar charMap of
              Just prevIndex ->
                let newStart = max prevIndex start + 1
                 in go
                      newStart
                      (end + 1)
                      (max maxLength (end - newStart + 1))
                      (Map.insert currentChar end charMap)
              Nothing ->
                go
                  start
                  (end + 1)
                  (max maxLength (end - start + 1))
                  (Map.insert currentChar end charMap)
--
-- $> lengthOfLongestSubstring "pwwkew"
